File: 003_Longest_Substring_Without_Repeating_Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        h = {}
        longestLen = 0
        start = 0
        for i in range(0, len(s)):
            if s[i] in h:
                start = max(start, h[s[i]] + 1)
            h[s[i]] = i
            longestLen = max(longestLen, i - start + 1)
        return longestLen

File: 003_Longest_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> dic;
        int start = 0, maxLen = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (dic.find(s[i]) != dic.end())
                start = max(start, dic[s[i]] + 1);
            maxLen = max(maxLen, i - start + 1);
            dic[s[i]] = i;
        }
        return maxLen;
    }
};
